src/libsyntax/parse/parser.rs

```diff
@@ -2921,6 +2937,27 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn is_at_start_of_range_notation_rhs(&self) -> bool {
+        if self.token.can_begin_expr() {
+            // parse `for i in 1.. { }` as infinite loop, not as `for i in (1..{})`.
+            if self.token == token::OpenDelim(token::Brace) {
+                return !self.restrictions.contains(RESTRICTION_NO_STRUCT_LITERAL);
+            }
+
+            // `1..*i` is ambiguous between `1..(*i)` and `(1..)*(i)`.
+            // We pick the `1..(*i)` interpretation.
+
+            // `r==1..&&true` is ambiguous between `r==(1..(&&true))` and `(r==(1..))&&true`.
+            // We pick the latter interpretation.
```

Comment:
I am not sure this is right. The fact that we have to have a special case for it kind of makes me feel that my intutions here are correct, and that having a special case around `..&&` feels kind of wrong. For one thing, it seems a bit odd that `1..&x` is fine but `1..&&x` is not?

On the other hand, I can't see a good reason to ever do `..&&x` whereas your example kind of makes sense.